File: scripts/carousel/generate_carousel.js
#!/usr/bin/env node
/**
 * scripts/carousel/generate_carousel.js
 *
 * proposal-data.json → 카드캐러셀 이미지 세트 (4:5 비율, 인스타그램 캐러셀용)
 *
 * 출력: output/carousel/<campaign-id>/slide-01.png ~ slide-0N.png
 *
 * 슬라이드 구성:
 *   1. 표지: 브랜드명 + 제품명 + "광고 전략 제안서"
 *   2. 전략 개요: strategy.overview
 *   3. 타겟 분석: strategy.targetAudience
 *   4. 핵심 메시지: strategy.keyMessage
 *   5. 차별화 포인트: strategy.differentiation
 *   6. 적용된 마케팅 원칙 (rationale 중 일부)
 *   7. 대본 미리보기 (script.scenes 중 주요 장면)
 *   8. CTA / 기대효과
 *
 * 환경: Node.js, Canvas (node-canvas) 또는 fallback으로 PNG 생성
 *       한글 폰트 필요.
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));

// ---------------------------------------------------------------------------
// 폰트 감지
// ---------------------------------------------------------------------------

function detectFont() {
  const candidates = [
    '/System/Library/Fonts/AppleSDGothicNeo.ttc',
    '/System/Library/Fonts/Supplemental/AppleGothic.ttf',
    '/Library/Fonts/AppleGothic.ttf',
  ];
  for (const f of candidates) {
    try {
      require('fs').accessSync(f);
      return f;
    } catch {}
  }
  return null;
}

// ---------------------------------------------------------------------------
// node-canvas 사용 시도
// ---------------------------------------------------------------------------

let canvasAPI = null;
try {
  const { createCanvas, loadImage, registerFont } = await import('canvas');
  const fontPath = detectFont();
  if (fontPath) {
    // .ttc는 canvas에서 직접 지원 안 될 수 있음 → 대안으로 AppleGothic.ttf 우선
    const fallback = '/System/Library/Fonts/Supplemental/AppleGothic.ttf';
    const useFont = require('fs').existsSync(fallback) ? fallback : fontPath;
    try {
      registerFont(useFont, { family: 'AppleGothic' });
    } catch {
      // registerFont 실패해도 진행
    }
    canvasAPI = { createCanvas, loadImage, fontFamily: 'AppleGothic', fontPath: useFont };
    console.log(`   canvas 폰트: ${useFont}`);
  }
} catch (e) {
  console.log('   node-canvas 사용 불가, fallback 모드');
}

// ---------------------------------------------------------------------------
// fallback: 순수 Node.js로는 PNG 생성이 어려우므로 PNG 출력 생략 가능
// ---------------------------------------------------------------------------

function ensureDir(path) {
  const d = typeof path === 'string' ? path : path.pathname;
  if (!existsSync(d)) mkdirSync(d, { recursive: true });
  return d;
}

// ---------------------------------------------------------------------------
// 캐러셀 생성 (canvas API)
// ---------------------------------------------------------------------------

async function generateCarousel(proposalPath, outputDir) {
  const proposal = JSON.parse(readFileSync(proposalPath, 'utf-8'));
  const campaignId = proposal.campaignId || 'carousel';
  const outDir = join(outputDir || join(__dirname, '..', '..', 'output', 'carousel'), campaignId);
  ensureDir(outDir);

  const brand = proposal.brandName || '브랜드';
  const product = proposal.productName || '제품';
  const strategy = proposal.strategy || {};
  const script = proposal.script || {};
  const rationale = proposal.rationale || [];

  const slides = [];

  // 1. 표지
  slides.push({
    title: `${brand} ${product}`,
    subtitle: '광고 전략 제안서',
    accent: '#2563EB',
    type: 'cover',
  });

  // 2. 전략 개요
  if (strategy.overview) {
    slides.push({
      title: '전략 개요',
      body: strategy.overview.split('\n').filter(Boolean).slice(0, 5).join('\n'),
      type: 'content',
    });
  }

  // 3. 타겟 분석
  if (strategy.targetAudience) {
    slides.push({
      title: '타겟 분석',
      body: strategy.targetAudience,
      type: 'content',
    });
  }

  // 4. 핵심 메시지
  if (strategy.keyMessage) {
    slides.push({
      title: '핵심 메시지',
      body: strategy.keyMessage,
      type: 'content',
    });
  }

  // 5. 차별화 포인트
  if (strategy.differentiation) {
    slides.push({
      title: '차별화 포인트',
      body: strategy.differentiation,
      type: 'content',
    });
  }

  // 6. 적용된 원칙 (상위 4개)
  const topPrinciples = rationale.slice(0, 4);
  if (topPrinciples.length > 0) {
    const body = topPrinciples.map((r, i) => {
      const name = r.principleName || r.principleId || '';
      const reason = (r.reason || '').substring(0, 100);
      return `${i + 1}. ${name}\n${reason}`;
    }).join('\n\n');
    slides.push({
      title: `적용된 마케팅 원칙 (${topPrinciples.length}개)`,
      body,
      type: 'content',
    });
  }

  // 7. 대본 미리보기 (주요 3장면)
  const scenes = script.scenes || [];
  const previewScenes = scenes.filter(s => s.type === 'hook').slice(0, 1)
    .concat(scenes.filter(s => s.type === 'solution').slice(0, 1))
    .concat(scenes.filter(s => s.type === 'cta').slice(0, 1));
  if (previewScenes.length > 0) {
    const body = previewScenes.map(s => {
      const typeLabel = s.type ? s.type.toUpperCase() : '';
      return `[${typeLabel}] ${s.dialogue || ''}`;
    }).join('\n\n');
    slides.push({
      title: '대본 미리보기',
      body: body || '(대본 없음)',
      type: 'content',
    });
  }

  // 8. CTA / 기대효과
  slides.push({
    title: '지금 바로시작하세요',
    body: `👉 ${brand} ${product} 자세히 보기`,
    type: 'cta',
  });

  // 렌더링
  if (!canvasAPI) {
    console.log('   ⚠️ node-canvas 없음 — 캐러셀 이미지 생성 건너뜀');
    return { slides: slides.map((s, i) => ({ index: i + 1, path: null, title: s.title })) };
  }

  const { createCanvas, fontFamily } = canvasAPI;
  const canvasW = 1080;
  const canvasH = 1350; // 4:5 비율

  const results = [];

  for (let i = 0; i < slides.length; i++) {
    const slide = slides[i];
    const canvas = createCanvas(canvasW, canvasH);
    const ctx = canvas.getContext('2d');

    // 배경
    const grad = ctx.createLinearGradient(0, 0, 0, canvasH);
    if (slide.type === 'cover') {
      grad.addColorStop(0, '#1E3A5F');
      grad.addColorStop(1, '#0D1B2A');
    } else if (slide.type === 'cta') {
      grad.addColorStop(0, '#2563EB');
      grad.addColorStop(1, '#1E40AF');
    } else {
      grad.addColorStop(0, '#F8FAFC');
      grad.addColorStop(1, '#E2E8F0');
    }
    ctx.fillStyle = grad;
    ctx.fillRect(0, 0, canvasW, canvasH);

    // 폰트 설정 (canvas font 문자열)
    const fontSizeTitle = slide.type === 'cover' ? 72 : 48;
    const fontSizeBody = 32;
    const fontSizeAccent = 28;

    ctx.textAlign = 'center';
    ctx.textBaseline = 'middle';

    // 표지 슬라이드
    if (slide.type === 'cover') {
      // 제목
      ctx.fillStyle = '#FFFFFF';
      ctx.font = `bold ${fontSizeTitle}px "${fontFamily}"`;
      wrapText(ctx, slide.title, canvasW * 0.85, canvasH * 0.38, fontSizeTitle);

      // 부제목
      ctx.fillStyle = 'rgba(255,255,255,0.8)';
      ctx.font = `500 ${fontSizeAccent}px "${fontFamily}"`;
      wrapText(ctx, slide.subtitle, canvasW * 0.85, canvasH * 0.52, fontSizeAccent);

      // 구분선
      ctx.strokeStyle = 'rgba(255,255,255,0.3)';
      ctx.lineWidth = 2;
      ctx.beginPath();
      ctx.moveTo(canvasW * 0.2, canvasH * 0.58);
      ctx.lineTo(canvasW * 0.8, canvasH * 0.58);
      ctx.stroke();

      // 페이지 번호
      ctx.fillStyle = 'rgba(255,255,255,0.5)';
      ctx.font = `400 24px "${fontFamily}"`;
      ctx.fillText(`${i + 1} / ${slides.length}`, canvasW / 2, canvasH - 60);
    } else if (slide.type === 'cta') {
      // CTA
      ctx.fillStyle = '#FFFFFF';
      ctx.font = `bold ${fontSizeTitle}px "${fontFamily}"`;
      wrapText(ctx, slide.title, canvasW * 0.85, canvasH * 0.38, fontSizeTitle);

      ctx.fillStyle = 'rgba(255,255,255,0.9)';
      ctx.font = `500 ${fontSizeBody}px "${fontFamily}"`;
      wrapText(ctx, slide.body, canvasW * 0.85, canvasH * 0.52, fontSizeBody);

      // 페이지 번호
      ctx.fillStyle = 'rgba(255,255,255,0.6)';
      ctx.font = `400 24px "${fontFamily}"`;
      ctx.fillText(`${i + 1} / ${slides.length}`, canvasW / 2, canvasH - 60);
    } else {
      // 콘텐츠 슬라이드
      // 제목
      ctx.fillStyle = '#0F172A';
      ctx.font = `bold ${fontSizeTitle}px "${fontFamily}"`;
      wrapText(ctx, slide.title, canvasW * 0.9, canvasH * 0.14, fontSizeTitle);

      // 구분선
      ctx.strokeStyle = '#CBD5E1';
      ctx.lineWidth = 2;
      ctx.beginPath();
      ctx.moveTo(canvasW * 0.1, canvasH * 0.20);
      ctx.lineTo(canvasW * 0.9, canvasH * 0.20);
      ctx.stroke();

      // 본문 (여러 줄)
      ctx.fillStyle = '#334155';
      ctx.font = `400 ${fontSizeBody}px "${fontFamily}"`;
      wrapText(ctx, slide.body, canvasW * 0.85, canvasH * 0.28, fontSizeBody);

      // 페이지 번호
      ctx.fillStyle = '#94A3B8';
      ctx.font = `400 24px "${fontFamily}"`;
      ctx.fillText(`${i + 1} / ${slides.length}`, canvasW / 2, canvasH - 50);
    }

    // 저장
    const outPath = join(outDir, `slide-${String(i + 1).padStart(2, '0')}.png`);
    const buffer = canvas.toBuffer('image/png');
    writeFileSync(outPath, buffer);
    results.push({ index: i + 1, path: outPath, title: slide.title });
    console.log(`   slide ${i + 1}: ${slide.title.substring(0, 30)} → ${outPath}`);
  }

  return { slides: results, count: slides.length };
}

// ---------------------------------------------------------------------------
// 텍스트 랩핑 헬퍼 (canvas context용)
// ---------------------------------------------------------------------------

function wrapText(ctx, text, maxWidth, y, fontSize) {
  const lines = text.split('\n');
  let currentY = y;
  for (const line of lines) {
    const words = line.split('');
    let lineText = '';
    for (const ch of words) {
      const test = lineText + ch;
      const metrics = ctx.measureText(test);
      if (metrics.width > maxWidth && lineText) {
        ctx.fillText(lineText, ctx.canvas.width / 2, currentY);
        currentY += fontSize * 1.3;
        lineText = ch;
      } else {
        lineText = test;
      }
    }
    if (lineText) {
      ctx.fillText(lineText, ctx.canvas.width / 2, currentY);
      currentY += fontSize * 1.3;
    }
  }
}

// ---------------------------------------------------------------------------
// 진입점
// ---------------------------------------------------------------------------

const proposalArg = process.argv[2];
const outputArg = process.argv[3];

if (!proposalArg) {
  console.log('사용법: node scripts/carousel/generate_carousel.js <proposal-data.json> [output_dir]');
  console.log('예: node scripts/carousel/generate_carousel.js content/campaigns/real-전자기기-xxx/proposal-data.json');
  process.exit(1);
}

(async () => {
  try {
    console.log(`📱 캐러셀 생성 시작: ${proposalArg}`);
    const result = await generateCarousel(proposalArg, outputArg);
    console.log(`\n✅ 캐러셀 생성 완료: ${result.count}장`);
    for (const s of result.slides) {
      console.log(`   slide ${s.index}: ${s.path || '생성 안 됨'}`);
    }
    process.exit(0);
  } catch (err) {
    console.error('❌ 캐러셀 생성 실패:', err.message);
    console.error(err.stack?.substring(0, 500));
    process.exit(1);
  }
})();

File: scripts/carousel/generate_carousel.py
#!/usr/bin/env python3
"""
scripts/carousel/generate_carousel.py
proposal-data.json → 카드캐러셀 이미지 세트 (4:5, 인스타그램 캐러셀용)

출력: output/carousel/<campaign-id>/slide-01.png ~ slide-0N.png

슬라이드 구성:
  1. 표지: 브랜드명 + 제품명 + "광고 전략 제안서"
  2. 전략 개요
  3. 타겟 분석
  4. 핵심 메시지
  5. 차별화 포인트
  6. 적용된 마케팅 원칙
  7. 대본 미리보기
  8. CTA
"""

import sys
import json
import os
from pathlib import Path
from PIL import Image, ImageDraw, ImageFont

# ---------------------------------------------------------------------------
# 폰트 감지
# ---------------------------------------------------------------------------

def detect_font():
    candidates = [
        '/System/Library/Fonts/AppleSDGothicNeo.ttc',
        '/System/Library/Fonts/Supplemental/AppleGothic.ttf',
        '/Library/Fonts/AppleGothic.ttf',
        '/usr/share/fonts/truetype/nanum/NanumGothic.ttf',
    ]
    for fp in candidates:
        if os.path.exists(fp):
            return fp
    return None

# ---------------------------------------------------------------------------
# 텍스트 줄바꿈
# ---------------------------------------------------------------------------

def wrap_text(draw, text, font, max_width):
    lines = []
    for paragraph in text.split('\n'):
        if not paragraph.strip():
            lines.append('')
            continue
        words = list(paragraph)
        current = ''
        for ch in words:
            test = current + ch
            bbox = draw.textbbox((0, 0), test, font=font)
            w = bbox[2] - bbox[0]
            if w > max_width and current:
                lines.append(current)
                current = ch
            else:
                current = test
        if current:
            lines.append(current)
    return lines

# ---------------------------------------------------------------------------
# 슬라이드 렌더링
# ---------------------------------------------------------------------------

def render_slide(slide_num, total, slide, font_path, out_path):
    W, H = 1080, 1350  # 4:5
    img = Image.new('RGB', (W, H), color=(20, 20, 30))
    draw = ImageDraw.Draw(img)

    font_title = ImageFont.truetype(font_path, 64)
    font_subtitle = ImageFont.truetype(font_path, 36)
    font_body = ImageFont.truetype(font_path, 32)
    font_page = ImageFont.truetype(font_path, 24)
    font_small = ImageFont.truetype(font_path, 22)

    title = slide.get('title', '')
    body = slide.get('body', '')
    slide_type = slide.get('type', 'content')
    accent_color = slide.get('accent', (37, 99, 235))

    if slide_type == 'cover':
        # 배경 그라데이션 (단색 + 강조색으로 대체)
        bg_color = (13, 27, 42)  # dark navy
        img.paste((bg_color[0], bg_color[1], bg_color[2]), (0, 0, W, H))

        # 제목
        draw.text((W//2, H*2//5), title, fill=(255, 255, 255), font=font_title, anchor='mm')
        # 부제목
        draw.text((W//2, H*3//5), slide.get('subtitle', ''), fill=(200, 200, 220), font=font_subtitle, anchor='mm')
        # 구분선
        draw.line([(W*0.2, H*0.62), (W*0.8, H*0.62)], fill=(100, 100, 140), width=2)
        # 페이지 번호
        draw.text((W//2, H-50), f'{slide_num} / {total}', fill=(150, 150, 170), font=font_page, anchor='mm')

    elif slide_type == 'cta':
        # CTA 배경
        img.paste((30, 58, 166), (0, 0, W, H))  # blue

        draw.text((W//2, H*3//8), title, fill=(255, 255, 255), font=font_title, anchor='mm')
        if body:
            draw.text((W//2, H*5//8), body, fill=(220, 230, 250), font=font_body, anchor='mm')
        draw.text((W//2, H-50), f'{slide_num} / {total}', fill=(180, 200, 230), font=font_page, anchor='mm')

    else:
        # 콘텐츠 슬라이드
        # 배경: 밝은 톤
        img.paste((248, 250, 252), (0, 0, W, H))

        # 제목
        draw.text((W//2, H*2//15), title, fill=(15, 23, 42), font=font_title, anchor='mm')

        # 구분선
        draw.line([(W*0.1, H*3//15), (W*0.9, H*3//15)], fill=(203, 213, 225), width=2)

        # 본문
        if body:
            body_lines = wrap_text(draw, body, font_body, W * 0.82)
            y_start = H * 4 // 15
            for i, line in enumerate(body_lines):
                draw.text((W//2, y_start + i * 48), line, fill=(71, 85, 105), font=font_body, anchor='mm')

        # 페이지 번호
        draw.text((W//2, H-45), f'{slide_num} / {total}', fill=(148, 163, 184), font=font_page, anchor='mm')

    img.save(out_path, 'PNG')
    return out_path

# ---------------------------------------------------------------------------
# 메인
# ---------------------------------------------------------------------------

def main():
    if len(sys.argv) < 2:
        print('사용법: python scripts/carousel/generate_carousel.py <proposal-data.json> [output_dir]')
        sys.exit(1)

    proposal_path = Path(sys.argv[1])
    output_dir = Path(sys.argv[2]) if len(sys.argv) > 2 else Path(__file__).parent.parent.parent / 'output' / 'carousel'

    if not proposal_path.exists():
        print(f'❌ proposal-data.json 없음: {proposal_path}')
        sys.exit(1)

    font_path = detect_font()
    if not font_path:
        print('❌ 한글 폰트를 찾을 수 없습니다.')
        sys.exit(1)

    print(f'📱 캐러셀 생성 시작: {proposal_path}')
    print(f'   폰트: {font_path}')

    with open(proposal_path, 'r', encoding='utf-8') as f:
        proposal = json.load(f)

    campaign_id = proposal.get('campaignId', 'carousel')
    brand = proposal.get('brandName', '브랜드')
    product = proposal.get('productName', '제품')
    strategy = proposal.get('strategy', {})
    script = proposal.get('script', {})
    rationale = proposal.get('rationale', [])

    slides = []

    # 1. 표지
    slides.append({
        'title': f'{brand} {product}',
        'subtitle': '광고 전략 제안서',
        'type': 'cover',
    })

    # 2. 전략 개요
    overview = strategy.get('overview', '')
    if overview:
        slides.append({
            'title': '전략 개요',
            'body': overview,
            'type': 'content',
        })

    # 3. 타겟 분석
    target = strategy.get('targetAudience', '')
    if target:
        slides.append({
            'title': '타겟 분석',
            'body': target,
            'type': 'content',
        })

    # 4. 핵심 메시지
    key_msg = strategy.get('keyMessage', '')
    if key_msg:
        slides.append({
            'title': '핵심 메시지',
            'body': key_msg,
            'type': 'content',
        })

    # 5. 차별화 포인트
    diff = strategy.get('differentiation', '')
    if diff:
        slides.append({
            'title': '차별화 포인트',
            'body': diff,
            'type': 'content',
        })

    # 6. 적용된 원칙 (상위 4개)
    top_principles = rationale[:4]
    if top_principles:
        body_lines = []
        for i, r in enumerate(top_principles, 1):
            name = r.get('principleName', r.get('principleId', ''))
            reason = (r.get('reason') or '')[:120]
            body_lines.append(f'{i}. {name}\n{reason}')
        slides.append({
            'title': f'적용된 마케팅 원칙 ({len(top_principles)}개)',
            'body': '\n\n'.join(body_lines),
            'type': 'content',
        })

    # 7. 대본 미리보기 (hook + solution + cta)
    scenes = script.get('scenes', [])
    preview_types = ['hook', 'solution', 'cta']
    preview = []
    for pt in preview_types:
        found = next((s for s in scenes if s.get('type') == pt), None)
        if found:
            preview.append(f"[{pt.upper()}] {found.get('dialogue', '')}")
    if preview:
        slides.append({
            'title': '대본 미리보기',
            'body': '\n\n'.join(preview),
            'type': 'content',
        })

    # 8. CTA
    slides.append({
        'title': '지금 바로 시작하세요',
        'body': f'👉 {brand} {product} 자세히 보기',
        'type': 'cta',
    })

    # 렌더링
    out_dir = output_dir / campaign_id
    out_dir.mkdir(parents=True, exist_ok=True)

    print(f'   슬라이드 {len(slides)}장 렌더링...')
    for i, slide in enumerate(slides, 1):
        out_path = out_dir / f'slide-{i:02d}.png'
        render_slide(i, len(slides), slide, font_path, str(out_path))
        print(f'   ✅ slide {i:02d}: {slide["title"][:30]}')

    print(f'\n✅ 캐러셀 생성 완료: {len(slides)}장 → {out_dir}')

if __name__ == '__main__':
    main()
